### PR TITLE
Configure app to cache episodes when playing

### PR DESCRIPTION
**Pull request by mebarbosa**
## Description
- This is part of the Playback Skipping project: Prevent the playback position from skipping while listening to a streaming episode by fully downloading it
- The goal of this PR is to prevent the unexpected playback skipping that our users reported. According to some investigations we did in the past, we believe this is related to ads insertion so our approach to fix this is to implement a cache system by triggering a download when the user begins streaming an episode switching from the streaming episode to the downloaded episode once the download completes.
- Project thread internal reference: `pdeCcb-4ux-p2`
- Playback skipping section from: `pdeCcb-4gS-p2`
- Slack thread: `p1708533677908569-slack-C05RR9P9RAT`
- #1865

> [!note]
> While you are testing, you might encounter the existing issue https://github.com/Automattic/pocket-casts-android/issues/86 related to the downloading percentage. I won't address it here as it's not within the scope of this task

> [!note]
> To test this PR I will to have the local `CACHE_PLAYING_EPISODE` feature flag enabled.

### Notes
- Have the feature flag enabled to be able to test it
- To see if the playback if switching from streaming to downloaded file I put a log to print the value of `episodeLocation`. I will noticed that when you tap to play an episode it will print the URL of the episode first and then the path to the downloaded episode:

https://github.com/Automattic/pocket-casts-android/blob/c24f8591be2c747d3e3acc04712de04a9f025a2e/modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/playback/LocalPlayer.kt#L184

### Play Episode
1. Play an episode.
2. Wait for it to finish downloading.
3. Ensure the app switches from streaming to downloading the file.
4. Go to Profile -> Downloads.
5. Confirm that the podcast is listed in the download section.
6. Play another episode.
7. The first episode should be deleted, and the current one playing starts downloading

### Queue
1. Play an episode.
2. Wait for it to finish downloading.
3. Ensure the app switches from streaming to downloading the file.
4. Go to Profile -> Downloads.
5. Confirm that the podcast is listed in the download section.
6. Add some episodes to the queue.
7. Seek the currently playing episode to the end and wait for the next one from the queue to start playing.
8. When the next episode starts playing, the previous episode should be removed from the downloads screen (cache), and the current episode should start downloading.
9. Add more episodes to the Up Next queue.
10. Go to Profile -> Settings -> General -> Toggle "Play up next episode on tap" ON.
11. Go to the Up Next queue.
12. Tap on some episode from the queue.
13. The previous episode should be removed from the cache, and the current one should start downloading.

### Auto Play
1. Play an episode.
2. Wait for it to finish downloading.
3. Ensure the app switches from streaming to downloading the file.
4. Go to Profile -> Settings -> General -> Toggle Auto play ON.
5. Wait for the current episode to finish playing.
6. When the next episode from the same podcast starts playing, it should be added to the download queue.

### Mark as Played
1. Play an episode.
2. Wait for it to finish downloading.
3. Ensure the app switches from streaming to downloading the file.
4. Go to Profile -> Downloads.
5. Confirm that the podcast is listed in the download section.
6. Close the app.
7. Open the app.
8. Open Now Playing.
9. Click on the three dots at the bottom and then mark as played (Auto play toggle must be off for this test).
10. The cached episode should be deleted from the downloads screen.

### Manual Downloads vs Automatic Downloads
1. Play an episode.
2. Wait for it to finish downloading.
3. Ensure the app switches from streaming to downloading the file.
4. Manually download some episodes.
5. Go to Profile -> Downloads.
6. Play the manually downloaded episodes, switching between them, and make sure they are not deleted.

### Skip Last
1. Go to Profile -> Settings -> General -> Toggle Auto play ON.
2. Go to some Podcast settings and set the skip last.
3. Play an episode from this podcast you configured.
4. Wait for it to finish downloading.
5. Ensure the app switches from streaming to downloading the file.
6. Wait for the podcast to finish.
7. A new episode from the same podcast should start playing and downloading, and the previous one should be removed from the cache.

### Charging Phone
1. Go to Profile -> Settings -> Auto download -> Toggle "Only when charging" ON.
2. With a device not charging, try to play an episode.
3. Ensure that no episode is added to the cache.
4. Set the battery device state to charging.
5. Try to play an episode.
6. Ensure that the episode is added to the cache.

### Only on Unmetered WiFi
1. Go to Profile -> Settings -> Auto download -> Toggle "Only on unmetered WiFi" ON.
2. With a device not connected via WiFi, try to play an episode.
3. Ensure that no episode is added to the cache.
4. Connect to WiFi.
5. Try to play an episode.
6. Ensure that the episode is added to the cache.

### Automotive and Watch
1. Play an episode in Automotive and make sure the episode it’s not been cached since we don’t support downloads for Automotive
2. Play an episode in Watch and make sure the episode it’s not been cached

### Casting
1. Cast an episode
2. Make sure it still works as expected

## Checklist
- [ ] If this is a user-facing change, I have added an entry in CHANGELOG.md
- [ ] Ensure the linter passes (`./gradlew spotlessApply` to automatically apply formatting/linting)
- [ ] I have considered whether it makes sense to add tests for my changes
- [ ] All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`
- [ ] Any jetpack compose components I added or changed are covered by compose previews
 
#### I have tested any UI changes...

- [ ] with different themes
- [ ] with a landscape orientation
- [ ] with the device set to have a large display and font size
- [ ] for accessibility with TalkBack

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by CookieyedCodes** [comment]
Could we have as part of this project a setting that could download a set number of upnext? For both watch and the phone? (like the apple watch setting)

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/podcast/EpisodeManagerImpl.kt [hunk 1]
```diff
@@ -355,6 +355,24 @@ class EpisodeManagerImpl @Inject constructor(
         }
     }
 
+    override suspend fun updateAutomaticallyCachedStatus(
+        episode: BaseEpisode?,
+        automaticallyCached: Boolean,
+    ) {
+        if (episode == null || episode is UserEpisode) return
```
Comment (MiSikora) [inline modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/podcast/EpisodeManagerImpl.kt:362]: I find it extremely strange to accept `BaseEpisode?` when the only type we can work with is `UserEpisode`. I know this is a convention in the project, where we accept as inputs overly broad types, and I'm not requesting a change here. Just something that stands out to me.

**Review by MiSikora — COMMENTED** [review]
The feature works as described by I have a mixed feeling about the UX. Often switching from a stream to a downloaded file resulted in the playback jumping a second or two back. Is this something that you plan working on in the subsequent PRs?

**Comment by mebarbosa** [comment]
> Could we have as part of this project a setting that could download a set number of upnext? For both watch and the phone? (like the apple watch setting)

Hey @CookieyedCodes! Thanks for sharing your suggestion. It's not part of the scope right now, but I will take note of this.

**Comment by CookieyedCodes** [comment]
No bothers, just something I've wanted for a long time to cut down on the 30 GBs of pocketcasts storage I have 😅

**Comment by mebarbosa** [comment]
> Is this something that you plan working on in the subsequent PRs?

@MiSikora Thanks for testing!
If we take a look at this https://github.com/Automattic/pocket-casts-android/pull/1510 we can see this is related to another [existing issue](https://github.com/Automattic/pocket-casts-android/issues/1509):

> Observe that once the episode is downloaded, if the player gets reset (which does not always happen, see https://github.com/Automattic/pocket-casts-android/issues/1509), the audio will pause for ~1 second and then resume as the player gets reset, and the audio should NOT jump backwards by ~1 second (i.e., there should not hear any audio repeated both before the pause and after the pause). Note that you can check the logs for "Resetting player" to confirm whether the player was reset.

Maybe we can think to address this in another PR. Since this is behind a FF we are good. What do you think?

**Review by MiSikora — APPROVED** [review]
Yeah I don't mind fixing it in another PR at all. Just wanted to check if it is something planned because now with this feature it will much more prominent.

**Pull request merged**